# **Inteligencia Artificial con Aplicaciones en Economía I**

- 👩‍🏫 **Profesora:** [Lina María Castro](https://www.linkedin.com/in/lina-maria-castro)  
- 📧 **Email:** [lmcastroco@gmail.com](mailto:lmcastroco@gmail.com)  
- 🎓 **Universidad:** Universidad Externado de Colombia - Facultad de Economía

# **Parte práctica parcial: Preparación y limpieza de datos**🧹

**IMPORTANTE**: Guarda una copia de este notebook en tu Google Drive o computador.

**Nombre estudiante:** Carlos Andrés Jiménez Sánchez

**Forma de entrega**

Jupyter Notebook publicado en su cuenta de Github con el nombre “Practica_limpieza_nombre_apellido_estudiante.ipynb”. Máximo a las 9 a.m., debes enviar link del notebook al correo lina.castro6@uexternado.edu.co, de lo contrario, no será tenido en cuenta.

**Instrucciones Generales**

- Puedes utilizar tus apuntes y notebooks y entrar al repositorio del curso. Sin embargo, **NO puedes utilizar IA para desarrollar tus respuestas**. **En caso de descubrir que estás usando IA** para el desarrollo de los ejercicios, **tu nota** de este segundo parcial **será cero**.
- Completa el código en las celdas marcadas con `### TU CÓDIGO AQUÍ ###`. Puedes añadir más celdas si lo requieres.
- Después de cada punto, ejecuta la celda de `assert` que se proporciona. Si no aparece ningún error, ¡tu respuesta es correcta!
- No modifiques las celdas de `assert` ni las celdas de Markdown.

**Objetivos de Aprendizaje**

1. **Aplicar técnicas de importación y consolidación de múltiples fuentes de datos** utilizando la librería `pandas`.
2.  **Ejecutar un proceso completo de limpieza de datos**, incluyendo el manejo de valores nulos, duplicados y la estandarización de tipos de datos.
3.  **Preparar un conjunto de datos real (GEIH del DANE)** para un análisis económico, sentando las bases para futuros modelos de machine learning.

# Ejercicio 1: Importar y unir datos (20%)

El DANE publica los microdatos de la GEIH mensualmente. Tu primera tarea es cargar los datos de los primeros seis meses de 2025 y armar una sola base.

**Los archivos necesarios los encuentras en el repositorio del curso, carpeta "datasets", carpeta "geih_2025".**

1. Usa un ciclo "for" para abrir los archivos CSV, ten en cuenta que estas bases usan como separador ";" y encoding='latin1'. La variable 'DPTO' y 'RAMA4D_R4' debes cargarlas como string.

In [392]:
### TU CÓDIGO AQUÍ ###

import pandas as pd
import numpy as np
import os

# Que muestre todas las columnas
pd.options.display.max_columns = None
# En los dataframes, mostrar los float con dos decimales
pd.options.display.float_format = '{:,.2f}'.format

from google.colab import drive, files
drive.mount('/content/drive')

ruta_csv = '/content/drive/MyDrive/geih_2025'

lista_archivos= os.listdir(ruta_csv)

dfs = []

for archivo in lista_archivos:
    csvs_ruta = os.path.join(ruta_csv, archivo)
    if archivo.endswith(".csv"):
        df_cvs = pd.read_csv(csvs_ruta, sep=';', encoding='latin1', dtype={'DPTO': str, 'RAMA4D_R4': str})
        dfs.append(df_cvs)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


/tmp/ipython-input-3520540514.py:24: DtypeWarning: Columns (71,132,176,185,190,201) have mixed types. Specify dtype option on import or set low_memory=False.
  df_cvs = pd.read_csv(csvs_ruta, sep=';', encoding='latin1', dtype={'DPTO': str, 'RAMA4D_R4': str})
/tmp/ipython-input-3520540514.py:24: DtypeWarning: Columns (71,132,176,188,190,201) have mixed types. Specify dtype option on import or set low_memory=False.
  df_cvs = pd.read_csv(csvs_ruta, sep=';', encoding='latin1', dtype={'DPTO': str, 'RAMA4D_R4': str})
/tmp/ipython-input-3520540514.py:24: DtypeWarning: Columns (71,132,176,188,190,201) have mixed types. Specify dtype option on import or set low_memory=False.
  df_cvs = pd.read_csv(csvs_ruta, sep=';', encoding='latin1', dtype={'DPTO': str, 'RAMA4D_R4': str})
/tmp/ipython-input-3520540514.py:24: DtypeWarning: Columns (71,132,176,188,190,201) have mixed types. Specify dtype option on import or set low_memory=False.
  df_cvs = pd.read_csv(csvs_ruta, sep=';', encoding='latin1', dty

2. Une todos los dataframes para obtener una base de la GEIH con toda la información de enero a junio de 2025. El nuevo dataframe debe llamarse `df_geih`

In [393]:
### TU CÓDIGO AQUÍ ###
df_geih = pd.concat(dfs, axis=0, ignore_index=True)

3. Reinicia el índice de `df_geih`. Asegúrate de que el índice antiguo no se añada como una nueva columna.

In [394]:
### TU CÓDIGO AQUÍ ###
df_geih.reset_index(drop=True, inplace=True)
display(df_geih)

PERIODO  MES   PER  DIRECTORIO  SECUENCIA_P  ORDEN  HOGAR  REGIS  \
0       20250105    1  2025     8086872            1      1      1     10   
1       20250105    1  2025     8086874            1      1      1     10   
2       20250105    1  2025     8086874            1      2      1     10   
3       20250105    1  2025     8086874            1      3      1     10   
4       20250105    1  2025     8086875            1      1      1     10   
...          ...  ...   ...         ...          ...    ...    ...    ...   
408418  20250518    5  2025     8238894            1      3      1     10   
408419  20250518    5  2025     8238894            1      4      1     10   
408420  20250518    5  2025     8238894            1      5      1     10   
408421  20250518    5  2025     8238894            1      6      1     10   
408422  20250518    5  2025     8238895            1      1      1     10   

        AREA  CLASE  FEX_C18 DPTO  PT  P6016  P3271  P6040  P6030S1  P6030S3  \
0        NaN      2 2,184.75   13   1      1      2     54     4.00 1,970.00   
1        NaN      2 1,208.49   13   1      1      2     79     3.00 1,945.00   
2        NaN      2 1,208.49   13   1      2      1     60     6.00 1,964.00   
3        NaN      2 1,208.49   13   1      3      1     59     2.00 1,965.00   
4        NaN      2 2,104.11   13   1      1      2     80     1.00 1,945.00   
...      ...    ...      ...  ...  ..    ...    ...    ...      ...      ...   
408418 66.00      1 1,871.24   66   1      1      2     14    11.00 2,010.00   
408419 66.00      1 1,871.24   66   1      1      2     11     6.00 2,014.00   
408420 66.00      1 1,871.24   66   1      1      1      9    10.00 2,015.00   
408421 66.00      1 1,871.24   66   1      6      1     70    10.00 1,954.00   
408422 66.00      1   576.95   66   1      1      1     22     9.00 2,002.00   

        P6050  P6083  P6083S1  P6081  P6081S1  P2057  P2059  P2061  P6080  \
0           1      3      NaN      3      NaN      1    NaN      1      6   
1           1      3      NaN      3      NaN      1    NaN      2      5   
2           9      3      NaN      3      NaN      1    NaN      2      5   
3           9      3      NaN      3      NaN      1    NaN      1      5   
4           1      2      NaN      3      NaN      2   2.00      2      5   
...       ...    ...      ...    ...      ...    ...    ...    ...    ...   
408418      3      1     1.00      2      NaN      2   2.00      2      6   
408419      3      1     1.00      1     2.00      2   2.00      2      6   
408420      3      1     1.00      1     2.00      2   2.00      2      6   
408421      5      3      NaN      3      NaN      2   2.00      2      6   
408422      1      2      NaN      2      NaN      1    NaN      2      6   

       P6080S1  P6080S1A1  P6070  P6071  P6071S1  P6090  P6100  P6110  \
0          NaN        NaN   4.00    NaN      NaN      1   3.00    NaN   
1          NaN        NaN   6.00    NaN      NaN      1   3.00    NaN   
2          NaN        NaN   6.00    NaN      NaN      1   3.00    NaN   
3          NaN        NaN   4.00    NaN      NaN      1   3.00    NaN   
4          NaN        NaN   4.00    NaN      NaN      1   2.00   2.00   
...        ...        ...    ...    ...      ...    ...    ...    ...   
408418     NaN        NaN   6.00    NaN      NaN      1   1.00   5.00   
408419     NaN        NaN   6.00    NaN      NaN      1   1.00   5.00   
408420     NaN        NaN    NaN    NaN      NaN      1   1.00   5.00   
408421     NaN        NaN   4.00    NaN      NaN      1   1.00   2.00   
408422     NaN        NaN   6.00    NaN      NaN      1   1.00   1.00   

            P6120  P1906S1  P1906S2  P1906S3  P1906S4  P1906S5  P1906S6  \
0             NaN        4        4        4        4        4        4   
1             NaN        4        4        4        4        4        4   
2             NaN        4        4        4        4        4        4   
3             NaN    

In [395]:
# Celda de verificación (no modificar)
assert df_geih.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_geih.shape[0]}."
assert df_geih.shape[1] == 245, f"Error en las columnas. Se esperaban 245, pero se encontraron {df_geih.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
assert 'index' not in df_geih.columns, "El índice antiguo no debe ser una columna."
assert df_geih.index.is_unique, "El índice debe ser único después de reiniciarlo."
print(f"¡Punto 1 completado con éxito! ✅ El DataFrame consolidado tiene {df_geih.shape[0]} filas y {df_geih.shape[1]} columnas.")

¡Las dimensiones del DataFrame son correctas! ✅
¡Punto 1 completado con éxito! ✅ El DataFrame consolidado tiene 408423 filas y 245 columnas.


# Ejercicio 2:  Seleccionar y renombrar columnas (5%)

La GEIH contiene cientos de variables. Para nuestro análisis del mercado laboral, nos enfocaremos en un subconjunto.

Adicionalmente, los nombres de las columnas en la GEIH son códigos. Para hacer el análisis más intuitivo, vamos a renombrarlas.

1. Crea un nuevo DataFrame llamado `df_empleo` que contenga únicamente las columnas que están en la lista `columnas_interes`.

In [396]:
columnas_interes = ['DIRECTORIO', 'SECUENCIA_P', 'ORDEN', 'MES', 'CLASE', 'DPTO', 'FEX_C18', 'PT', 'P3271', 'P6040', 'P6080', 'P3043', 'OCI', 'P6430', 'RAMA4D_R4', 'INGLABO']

In [397]:
### TU CÓDIGO AQUÍ ###
df_empleo = df_geih[columnas_interes]

In [398]:
# Celda de verificación (no modificar)
assert len(df_empleo.columns) == 16, "El número de columnas es incorrecto."
assert all(col in df_empleo.columns for col in columnas_interes), "Faltan columnas o los nombres son incorrectos."
print("¡Selección de columnas completada con éxito! ✅")
print("Dimensiones del nuevo DataFrame:", df_empleo.shape)

¡Selección de columnas completada con éxito! ✅
Dimensiones del nuevo DataFrame: (408423, 16)


2. Renombra las columnas del DataFrame `df_empleo` utilizando el diccionario `mapa_nombres`.

In [399]:
mapa_nombres = {
'DIRECTORIO': 'id_vivienda',
'SECUENCIA_P': 'id_hogar',
'ORDEN': 'id_persona',
'MES': 'mes',
'CLASE': 'clase',
'DPTO': 'departamento',
'FEX_C18': 'factor_expansion',
'PT': 'poblacion_total',
'P3271': 'sexo',
'P6040': 'edad',
'P6080': 'grupo_etnico',
'P3043': 'nivel_educativo',
'OCI': 'ocupados',
'P6430': 'categoria_ocupacional',
'RAMA4D_R4': 'cod_ciiu',
'INGLABO': 'ingreso_laboral'}

In [400]:
### TU CÓDIGO AQUÍ ###
df_empleo = df_empleo.rename(columns=mapa_nombres)

In [401]:
# Celda de verificación (no modificar)
nombres_nuevos_esperados = list(mapa_nombres.values())
condicion_exitosa = all(col in df_empleo.columns for col in nombres_nuevos_esperados)
assert condicion_exitosa, "No todas las columnas esperadas se encuentran en el DataFrame. Revisa el diccionario o la operación de renombrado."
print("¡Verificación exitosa! Todas las columnas fueron renombradas correctamente. ✅")
print("¡Punto 2 completado con éxito! ✅")

¡Verificación exitosa! Todas las columnas fueron renombradas correctamente. ✅
¡Punto 2 completado con éxito! ✅


# Ejercicio 3: Revisión de duplicados (5%)

Un paso crucial en la limpieza es asegurar que no haya observaciones duplicadas que puedan sesgar nuestro análisis.

1. Verifica si hay filas duplicadas en `df_empleo`.

In [402]:
### TU CÓDIGO AQUÍ ###
df_empleo[df_empleo.duplicated(keep=False)]

Empty DataFrame
Columns: [id_vivienda, id_hogar, id_persona, mes, clase, departamento, factor_expansion, poblacion_total, sexo, edad, grupo_etnico, nivel_educativo, ocupados, categoria_ocupacional, cod_ciiu, ingreso_laboral]
Index: []

2. Cuenta cuántas filas duplicadas existen.

In [403]:
### TU CÓDIGO AQUÍ ###
df_empleo.duplicated().sum()

np.int64(0)

3. Elimina las filas duplicadas, si las hay.

In [404]:
### TU CÓDIGO AQUÍ ###


4. Verifica si hay filas duplicadas específicamente en las columnas 'id_vivienda', 'id_hogar', 'id_persona'.

In [405]:
### TU CÓDIGO AQUÍ ###
df_empleo[df_empleo.duplicated(subset=['id_vivienda','id_hogar', 'id_persona'], keep=False)]

Empty DataFrame
Columns: [id_vivienda, id_hogar, id_persona, mes, clase, departamento, factor_expansion, poblacion_total, sexo, edad, grupo_etnico, nivel_educativo, ocupados, categoria_ocupacional, cod_ciiu, ingreso_laboral]
Index: []

In [406]:
# Celda de verificación (no modificar)
assert df_empleo.duplicated().sum() == 0, "Aún existen filas duplicadas."
print("¡Punto 3 completado con éxito! ✅")

¡Punto 3 completado con éxito! ✅


# Ejercicio 4: Valores faltantes (5%)

Los datos del mundo real raramente están completos. Aquí, trataremos los valores faltantes en la columna de ingresos.


1. Calcula el número de valores nulos (`NaN`) en la columna `ingreso_laboral` **antes** de cualquier modificación.

In [407]:
### TU CÓDIGO AQUÍ ###
df_empleo['ingreso_laboral'].isna().sum()

np.int64(236576)

2. En la GEIH, un ingreso faltante a menudo significa que la persona no reportó ingresos o no aplica (ej. no trabaja). Una estrategia común es imputar estos valores con `0`. Rellena los valores `NaN` en la columna `ingreso_laboral` con `0`.

In [408]:
### TU CÓDIGO AQUÍ ###
df_empleo['ingreso_laboral'] = df_empleo['ingreso_laboral'].fillna(0)
df_empleo

id_vivienda  id_hogar  id_persona  mes  clase departamento  \
0           8086872         1           1    1      2           13   
1           8086874         1           1    1      2           13   
2           8086874         1           2    1      2           13   
3           8086874         1           3    1      2           13   
4           8086875         1           1    1      2           13   
...             ...       ...         ...  ...    ...          ...   
408418      8238894         1           3    5      1           66   
408419      8238894         1           4    5      1           66   
408420      8238894         1           5    5      1           66   
408421      8238894         1           6    5      1           66   
408422      8238895         1           1    5      1           66   

        factor_expansion  poblacion_total  sexo  edad  grupo_etnico  \
0               2,184.75                1     2    54             6   
1               1,208.49                1     2    79             5   
2               1,208.49                1     1    60             5   
3               1,208.49                1     1    59             5   
4               2,104.11                1     2    80             5   
...                  ...              ...   ...   ...           ...   
408418          1,871.24                1     2    14             6   
408419          1,871.24                1     2    11             6   
408420          1,871.24                1     1     9             6   
408421          1,871.24                1     1    70             6   
408422            576.95                1     1    22             6   

        nivel_educativo  ocupados  categoria_ocupacional cod_ciiu  \
0                   NaN      1.00                   3.00     9700   
1                   NaN       NaN                    NaN      NaN   
2                   NaN      1.00                   4.00     0113   
3                   NaN      1.00                   1.00     0141   
4                  7.00       NaN                    NaN      NaN   
...                 ...       ...                    ...      ...   
408418              NaN       NaN                    NaN      NaN   
408419              NaN       NaN                    NaN      NaN   
408420              NaN       NaN                    NaN      NaN   
408421              NaN      1.00                   4.00     5511   
408422             5.00      1.00                   1.00     5224   

        ingreso_laboral  
0          1,040,000.00  
1                  0.00  
2            700,000.00  
3          1,450,000.00  
4                  0.00  
...                 ...  
408418             0.00  
408419             0.00  
408420             0.00  
408421     1,200,000.00  
408422     1,400,000.00  

[408423 rows x 16 columns]

In [409]:
# Celda de verificación (no modificar)
assert df_empleo['ingreso_laboral'].isnull().sum() == 0, "La columna 'ingreso_laboral' todavía tiene valores nulos."
print("¡Punto 4 completado con éxito! ✅")

¡Punto 4 completado con éxito! ✅


# Ejercicio 5: Decodificar variables (20%)



Las variables categóricas a menudo necesitan ser "traducidas" para hacerlas más entendibles para el análisis.

1. Usa los siguientes diccionarios para reemplazar los códigos numéricos por etiquetas de texto en las siguientes columnas: 'clase', 'sexo', 'grupo_etnico', 'nivel_educativo' y 'categoria_ocupacional'. Muestra los conteos de las nuevas categorías usando `.value_counts()` para verificar el cambio en cada una de las columnas.

In [410]:
dic_clase = {1: 'Cabeceras', 2: 'Centros poblados y rural disperso'}
dic_sexo = {1: 'Hombre', 2: 'Mujer'}
dic_etnico = {1: 'Indígena', 2: 'Gitano (Rom)', 3: 'Raizal del archipiélago', 4: 'Palenquero',
              5: 'Negro, mulato (afrodescendiente)', 6: 'Ninguno de los anteriores'}
dic_nivel_educativo = {1: 'Ninguno', 2: 'Educación media', 3: 'Educación media',
                       4: 'Educación media', 5: 'Técnica profesional', 6: 'Tecnológica',
                       7: 'Universitaria', 8: 'Especialización', 9: 'Maestría',
                       10: 'Doctorado', 99: 'No sabe, no informa'}
dic_cat_ocu = {1: 'Asalariado', 2: 'Asalariado', 3: 'Asalariado',
               8: 'Asalariado', 4: 'Independiente', 5: 'Independiente',
               9: 'Independiente', 6: 'Trabajador sin remuneración',
               7: 'Trabajador sin remuneración'}

In [411]:
### TU CÓDIGO AQUÍ ###
df_empleo["clase"] = df_empleo["clase"].replace(dic_clase)
df_empleo["sexo"] = df_empleo["sexo"].replace(dic_sexo)
df_empleo["grupo_etnico"] = df_empleo["grupo_etnico"].replace(dic_etnico)
df_empleo["nivel_educativo"] = df_empleo["nivel_educativo"].replace(dic_nivel_educativo)
df_empleo["categoria_ocupacional"] = df_empleo["categoria_ocupacional"].replace(dic_cat_ocu)

print(df_empleo["clase"].value_counts())
print(df_empleo["sexo"].value_counts())
print(df_empleo["grupo_etnico"].value_counts())
print(df_empleo["nivel_educativo"].value_counts())
print(df_empleo["categoria_ocupacional"].value_counts())

clase
Cabeceras                            356225
Centros poblados y rural disperso     52198
Name: count, dtype: int64
sexo
Mujer     218100
Hombre    190323
Name: count, dtype: int64
grupo_etnico
Ninguno de los anteriores           351537
Negro, mulato (afrodescendiente)     38516
Indígena                             17191
Raizal del archipiélago               1004
Palenquero                             126
Gitano (Rom)                            49
Name: count, dtype: int64
nivel_educativo
Educación media        117166
Universitaria           32763
Técnica profesional     24196
Ninguno                 12797
Tecnológica             10561
Especialización          7584
Maestría                 3067
Doctorado                 317
No sabe, no informa         3
Name: count, dtype: int64
categoria_ocupacional
Asalariado                     89162
Independiente                  80094
Trabajador sin remuneración     6920
Name: count, dtype: int64


In [412]:
# Celda de verificación (no modificar)
mapeos_a_verificar = {
    'clase': dic_clase,
    'sexo': dic_sexo,
    'grupo_etnico': dic_etnico,
    'nivel_educativo': dic_nivel_educativo,
    'categoria_ocupacional': dic_cat_ocu
}
for columna, diccionario in mapeos_a_verificar.items():
    es_numerica = pd.api.types.is_numeric_dtype(df_empleo[columna])
    assert not es_numerica, f"La columna '{columna}' todavía tiene un tipo de dato numérico. El reemplazo falló."
    valor_texto_ejemplo = list(diccionario.values())[0]
    categorias_unicas = df_empleo[columna].unique()
    assert valor_texto_ejemplo in categorias_unicas, f"No se encontró el valor '{valor_texto_ejemplo}' en la columna '{columna}'."
    valor_numerico_ejemplo = list(diccionario.keys())[0]
    assert valor_numerico_ejemplo not in categorias_unicas, f"El código numérico '{valor_numerico_ejemplo}' todavía existe en la columna '{columna}'."
print("¡Verificación completada con éxito! ✅ Todas las columnas categóricas fueron transformadas correctamente.")

¡Verificación completada con éxito! ✅ Todas las columnas categóricas fueron transformadas correctamente.


2. Utiliza el archivo "2025-01-14 Correlativa CIIU Rev4.xlsx", hoja 'CIIU 2022', para agregar al df_empleo las descripciones de los códigos CIIU a 4 dígitos. En df_empleo la columna con los códigos CIIU se llama 'cod_ciiu'. Si se agregan nuevas columnas al dataframe, distintas a la descripción del CIIU, debes borrarlas.

In [413]:
### TU CÓDIGO AQUÍ ###
ruta_correlativa_ciiu = '/content/drive/MyDrive/geih_2025/2025-01-14 Correlativa CIIU Rev4.xlsx'
df_ciiu = pd.read_excel(ruta_correlativa_ciiu, sheet_name='CIIU 2022')
df_ciiu['RAMA4D_R4'] = df_ciiu['RAMA4D_R4'].astype(str)
df_empleo = pd.merge(df_empleo, df_ciiu[['RAMA4D_R4', 'DESCRIPCION_CIIU']], left_on='cod_ciiu', right_on='RAMA4D_R4', how='left')
df_empleo = df_empleo.drop(columns=['RAMA4D_R4'])
df_empleo.sample(10)


id_vivienda  id_hogar  id_persona  mes  \
364260      8219178         1           4    5   
168799      8129403         1           1    2   
184489      8136015         1           5    2   
382199      8227820         1           1    5   
248892      8257656         1           1    6   
240046      8253939         1           2    6   
172582      8130909         1           4    2   
214735      8243456         1           3    6   
368907      8222143         1           1    5   
205110      8239374         1           1    6   

                                    clase departamento  factor_expansion  \
364260                          Cabeceras           66            412.80   
168799  Centros poblados y rural disperso           25          1,018.15   
184489                          Cabeceras           19          1,263.84   
382199                          Cabeceras           66            186.94   
248892                          Cabeceras           50            233.24   
240046  Centros poblados y rural disperso           20          1,793.17   
172582                          Cabeceras           54            600.35   
214735                          Cabeceras           68            436.85   
368907                          Cabeceras           11          3,409.99   
205110                          Cabeceras           05          1,399.21   

        poblacion_total    sexo  edad                      grupo_etnico  \
364260                1   Mujer    64         Ninguno de los anteriores   
168799                1   Mujer    48         Ninguno de los anteriores   
184489                1   Mujer    21  Negro, mulato (afrodescendiente)   
382199                1  Hombre    52         Ninguno de los anteriores   
248892                1  Hombre    62         Ninguno de los anteriores   
240046                1  Hombre    24         Ninguno de los anteriores   
172582                1   Mujer    21         Ninguno de los anteriores   
214735                1   Mujer    28         Ninguno de los anteriores   
368907                1   Mujer    35         Ninguno de los anteriores   
205110                1  Hombre    61         Ninguno de los anteriores   

        nivel_educativo  ocupados        categoria_ocupacional cod_ciiu  \
364260              NaN       NaN                          NaN      NaN   
168799              NaN       NaN                          NaN      NaN   
184489  Educación media       NaN                          NaN      NaN   
382199              NaN      1.00                   Asalariado     8544   
248892  Educación media       NaN                          NaN      NaN   
240046  Educación media      1.00  Trabajador sin remuneración     0141   
172582          Ninguno       NaN                          NaN      NaN   
214735  Educación media      1.00                Independiente     5619   
368907  Educación media      1.00                   Asalariado     5611   
205110              NaN      1.00                Independiente     8130   

        ingreso_laboral                                   DESCRIPCION_CIIU  
364260             0.00                                                NaN  
168799             0.00                                                NaN  
184489             0.00                                                NaN  
382199     1,623,500.00                         Educación de universidades  
248892             0.00                                                NaN  
240046     1,460,000.00                                                NaN  
172582             0.00                                                NaN  
214735       800,000.00  Otros tipos de expendio de comidas preparadas ...  
368907     1,000,000.00           Expendio a la mesa de comidas preparadas  
205110     2,500,000.00  Actividades de paisajismo y servicios de mante...

In [414]:
# Celda de verificación (no modificar)
assert 'DESCRIPCION_CIIU' in df_empleo.columns, \
    f"La columna '{'DESCRIPCION_CIIU'}' no se encontró. Revisa el nombre que le diste a la nueva columna."
assert df_empleo['DESCRIPCION_CIIU'].notna().sum() > 0, \
    f"La columna '{'DESCRIPCION_CIIU'}' está completamente vacía. Verifica las llaves del merge."
print("✅ Verificación del merge exitosa.")

✅ Verificación del merge exitosa.


3. Utiliza el archivo "2025-01-14 Divipola.xlsx", hoja 'Departamentos', para agregar al df_empleo los nombres de los departamentos. En df_empleo la columna con los códigos Divipola de los departamentos se llama 'departamento'. Si se agregan nuevas columnas al dataframe, distintas al nombre del departamento, debes borrarlas.

In [415]:
### TU CÓDIGO AQUÍ ###
ruta_divipola = '/content/drive/MyDrive/geih_2025/2025-01-14 DIVIPOLA.xlsx'

df_divipola = pd.read_excel(ruta_divipola, sheet_name="Departamentos", skiprows=9)
df_divipola['Código'] = df_divipola['Código'].astype(str)
df_empleo = pd.merge(df_empleo, df_divipola[["Código", "Nombre"]], left_on='departamento', right_on='Código', how='left')
df_empleo = df_empleo.drop(columns=['Código'])
df_empleo.sample(10)

id_vivienda  id_hogar  id_persona  mes      clase departamento  \
279029      8147294         1           2    3  Cabeceras           44   
250353      8258309         1           2    6  Cabeceras           63   
307959      8159181         1           2    3  Cabeceras           41   
352822      8214467         1           1    5  Cabeceras           47   
207112      8240172         1           2    6  Cabeceras           15   
371319      8223065         1           2    5  Cabeceras           68   
127964      8204408         1           1    4  Cabeceras           23   
336522      8170934         1           1    3  Cabeceras           41   
345280      8210124         1           4    5  Cabeceras           44   
233677      8251272         1           2    6  Cabeceras           70   

        factor_expansion  poblacion_total    sexo  edad  \
279029             49.88                1   Mujer    24   
250353            235.92                1   Mujer    30   
307959            927.41                1  Hombre    22   
352822            499.08                1   Mujer    63   
207112             50.39                1   Mujer    62   
371319            532.31                1   Mujer    37   
127964          2,704.12                1   Mujer    50   
336522            245.54                1   Mujer    62   
345280            105.60                1  Hombre     9   
233677             95.71                1   Mujer    80   

                     grupo_etnico  nivel_educativo  ocupados  \
279029  Ninguno de los anteriores  Educación media      1.00   
250353  Ninguno de los anteriores  Educación media      1.00   
307959  Ninguno de los anteriores  Educación media       NaN   
352822  Ninguno de los anteriores              NaN       NaN   
207112  Ninguno de los anteriores              NaN       NaN   
371319  Ninguno de los anteriores    Universitaria      1.00   
127964                   Indígena              NaN      1.00   
336522  Ninguno de los anteriores    Universitaria       NaN   
345280  Ninguno de los anteriores              NaN       NaN   
233677  Ninguno de los anteriores              NaN       NaN   

       categoria_ocupacional cod_ciiu  ingreso_laboral  \
279029         Independiente     9602       150,000.00   
250353            Asalariado     4771     1,423,500.00   
307959                   NaN      NaN             0.00   
352822                   NaN      NaN             0.00   
207112                   NaN      NaN             0.00   
371319            Asalariado     7020     2,200,000.00   
127964         Independiente     1690       390,000.00   
336522                   NaN      NaN             0.00   
345280                   NaN      NaN             0.00   
233677                   NaN      NaN             0.00   

                                         DESCRIPCION_CIIU      Nombre  
279029         Peluquería y otros tratamientos de belleza  LA GUAJIRA  
250353  Comercio al por menor de prendas de vestir y s...     QUINDÍO  
307959                                                NaN       HUILA  
352822                                                NaN   MAGDALENA  
207112                                                NaN      BOYACÁ  
371319              Actividades de consultoría de gestión   SANTANDER  
127964  Fabricación de otros productos de madera; fabr...     CÓRDOBA  
336522                                                NaN       HUILA  
345280                                                NaN  LA GUAJIRA  
233677                                                NaN       SUCRE

In [416]:
# Celda de verificación (no modificar)
assert 'Nombre' in df_empleo.columns, \
    f"La columna '{'Nombre'}' no se encontró. Revisa el nombre que le diste a la nueva columna."
assert df_empleo['Nombre'].notna().sum() > 0, \
    f"La columna '{'Nombre'}' está completamente vacía. Verifica las llaves del merge."
print("✅ Verificación del merge exitosa.")
assert df_empleo.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_empleo.shape[0]}."
assert df_empleo.shape[1] == 18, f"Error en las columnas. Se esperaban 18, pero se encontraron {df_empleo.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
print("\n¡Punto 5 completado con éxito! ✅")

✅ Verificación del merge exitosa.
¡Las dimensiones del DataFrame son correctas! ✅

¡Punto 5 completado con éxito! ✅


# Ejercicio 6: Crear nuevas variables y reorganizar columnas (10%)

La GEIH cuenta con un factor de expansión que permite extrapolar los datos de los encuestados a totales para la población colombiana. Usar este factor de expansión, permite calcular la población total del país, la población en edad de trabajar, el número de ocupados, el número de desempleados, entre otros datos.

1. Crea una nueva columna llamada "fex", la cual corresponderá a la columna 'factor_expansion' dividida en 6 (se divide en 6, ya que estamos tomando el dato de 6 meses).

In [417]:
### TU CÓDIGO AQUÍ ###
df_empleo["fex"]=(df_empleo["factor_expansion"]/6)

2. Crea una nueva columna llamada 'poblacion_total_ex', la cual corresponderá a la columna 'poblacion_total' multiplicada por la columna 'fex'.

In [418]:
### TU CÓDIGO AQUÍ ###
df_empleo["poblacion_total_ex"] = df_empleo["poblacion_total"] * df_empleo["fex"]

3. Crea una nueva columna llamada 'ocupados_ex', la cual corresponderá a la columna 'ocupados' multiplicada por la columna 'fex'.

In [419]:
### TU CÓDIGO AQUÍ ###
df_empleo["ocupados_ex"] = df_empleo["ocupados"]*df_empleo["fex"]

4. Crea una nueva columna llamada 'joven', la cual debe decir 'Población joven' si la persona encuestada tiene entre 15 y 28 años (15 y 28 incluidos). De lo contrario, debe decir 'Resto de población'.

In [438]:
### TU CÓDIGO AQUÍ ###
df_empleo["joven"]= np.where((df_empleo["edad"]>=15) & (df_empleo["edad"]<=28), "Población joven", "Resto de población")

5. Reorganiza las columnas con el siguiente orden: 'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase', 'departamento', 'Nombre', 'factor_expansion', 'fex', 'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven', 'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex', 'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU', 'ingreso_laboral'

In [440]:
### TU CÓDIGO AQUÍ ###
orden_deseado =  ['id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase', 'departamento', 'Nombre', 'factor_expansion', 'fex', 'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven', 'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex', 'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU', 'ingreso_laboral']
df_empleo = df_empleo[orden_deseado]

In [441]:
# Celda de verificación (no modificar)
orden_esperado = [
    'id_vivienda', 'id_hogar', 'id_persona', 'mes', 'clase',
    'departamento', 'Nombre', 'factor_expansion', 'fex',
    'poblacion_total', 'poblacion_total_ex', 'sexo', 'edad', 'joven',
    'grupo_etnico', 'nivel_educativo', 'ocupados', 'ocupados_ex',
    'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU', 'ingreso_laboral'
]
orden_actual = df_empleo.columns.tolist()
assert orden_actual == orden_esperado, \
    f"""El orden de las columnas es incorrecto.
    - Orden Esperado: {orden_esperado}
    - Orden Encontrado: {orden_actual}"""
assert df_empleo.shape[0] == 408423, f"Error en las filas. Se esperaban 408423, pero se encontraron {df_empleo.shape[0]}."
assert df_empleo.shape[1] == 22, f"Error en las columnas. Se esperaban 22, pero se encontraron {df_empleo.shape[1]}."
print("¡Las dimensiones del DataFrame son correctas! ✅")
print("\n¡Punto 6 completado con éxito! ✅")

¡Las dimensiones del DataFrame son correctas! ✅

¡Punto 6 completado con éxito! ✅


# Ejercicio 7: Cambiar tipo de dato (5%)



Para realizar cálculos y análisis correctamente, las columnas deben tener el tipo de dato adecuado (numérico, categórico, etc.).

1.  Muestra en pantalla los tipos de datos actuales de `df_empleo`.

In [448]:
### TU CÓDIGO AQUÍ ###
df_empleo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 408423 entries, 0 to 408422
Data columns (total 22 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   id_vivienda            408423 non-null  int64  
 1   id_hogar               408423 non-null  int64  
 2   id_persona             408423 non-null  int64  
 3   mes                    408423 non-null  int64  
 4   clase                  408423 non-null  object 
 5   departamento           408423 non-null  object 
 6   Nombre                 408423 non-null  object 
 7   factor_expansion       408423 non-null  float64
 8   fex                    408423 non-null  float64
 9   poblacion_total        408423 non-null  int64  
 10  poblacion_total_ex     408423 non-null  float64
 11  sexo                   408423 non-null  object 
 12  edad                   408423 non-null  int64  
 13  joven                  408423 non-null  object 
 14  grupo_etnico           408423 non-nu

2.  Las siguientes columnas representan categorías: 'clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico', 'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU'. Conviértelas al tipo `category` para optimizar el uso de memoria y facilitar análisis futuros.

In [453]:
### TU CÓDIGO AQUÍ ###
categorias= ['clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico', 'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu', 'DESCRIPCION_CIIU']
for columna in categorias:
    df_empleo[columna] = df_empleo[columna].astype('category')
df_empleo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 408423 entries, 0 to 408422
Data columns (total 22 columns):
 #   Column                 Non-Null Count   Dtype   
---  ------                 --------------   -----   
 0   id_vivienda            408423 non-null  int64   
 1   id_hogar               408423 non-null  int64   
 2   id_persona             408423 non-null  int64   
 3   mes                    408423 non-null  int64   
 4   clase                  408423 non-null  category
 5   departamento           408423 non-null  category
 6   Nombre                 408423 non-null  category
 7   factor_expansion       408423 non-null  float64 
 8   fex                    408423 non-null  float64 
 9   poblacion_total        408423 non-null  int64   
 10  poblacion_total_ex     408423 non-null  float64 
 11  sexo                   408423 non-null  category
 12  edad                   408423 non-null  int64   
 13  joven                  408423 non-null  category
 14  grupo_etnico        

In [452]:
# Celda de verificación (no modificar)
columnas_categoricas_esperadas = [
    'clase', 'departamento', 'Nombre', 'sexo', 'joven', 'grupo_etnico',
    'nivel_educativo', 'categoria_ocupacional', 'cod_ciiu',
    'DESCRIPCION_CIIU'
]
for columna in columnas_categoricas_esperadas:
    assert columna in df_empleo.columns, \
        f"La columna '{columna}' no fue encontrada en el DataFrame."
    es_categoria = pd.api.types.is_categorical_dtype(df_empleo[columna])
    assert es_categoria, \
        f"La columna '{columna}' no fue convertida al tipo 'category'. Su tipo actual es {df_empleo[columna].dtype}."
print("¡Verificación completada con éxito! ✅ Todas las columnas especificadas son del tipo 'category'.")
print("\n¡Punto 7 completado con éxito! ✅")

¡Verificación completada con éxito! ✅ Todas las columnas especificadas son del tipo 'category'.

¡Punto 7 completado con éxito! ✅


/tmp/ipython-input-416816491.py:10: DeprecationWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, pd.CategoricalDtype) instead
  es_categoria = pd.api.types.is_categorical_dtype(df_empleo[columna])


# Ejercicio 8: Ordenar el dataframe (5%)

1. Ordena el dataframe por la columna "edad" de mayor a menor. Observa la columna 'edad' y escribe en una celda de markdown cuál fue la edad mínima y la edad máxima de las personas que respondieron la encuesta de la GEIH.

In [471]:
### TU CÓDIGO AQUÍ ###
df_empleo = df_empleo.sort_values(by="edad", ascending=False)
df_empleo["edad"].describe()

count   408,423.00
mean         36.32
std          22.25
min           0.00
25%          18.00
50%          34.00
75%          54.00
max         106.00
Name: edad, dtype: float64

Escribe tu análisis aquí: Haciendo un analisis descriptivo de la variable edad se observa una edad maxima de 106 años y una minima de 0. Adicionalmente, la media de la muestra está en 36 años y la mediana en 34 años por lo que se podria decir que la distrubucion es aproximadamente simetrica en donde no se presenta un sesgo apreciable


In [470]:
# Celda de verificación (no modificar)
es_descendiente = df_empleo['edad'].is_monotonic_decreasing
assert es_descendiente, "La columna 'edad' no está ordenada correctamente de mayor a menor."
print("\n¡Punto 8 completado con éxito! ✅")


¡Punto 8 completado con éxito! ✅


# Ejercicio 9: Groupby (20%)

Agrupar los datos por categorías, nos permite realizar análisis más profundos y obtener insights valiosos, que a simple vista no vemos en la base de datos.

1. Crea un nuevo dataframe llamado "df_sexo" donde se agrupe por 'sexo' y se calcule la suma de las columnas 'poblacion_total_ex' y 'ocupados_ex'. Crea una nueva columna que se llame 'porcentaje_ocupados' donde dividas las columna 'ocupados_ex' sobre 'poblacion_total_ex'. Al final usa reset_index() para convertir las categorías de sexo en una columna.

In [489]:
### TU CÓDIGO AQUÍ ##
df_sexo= df_empleo.groupby("sexo")[["poblacion_total_ex","ocupados_ex"]].sum().reset_index()
df_sexo["porcentaje_ocupados"]= df_sexo["ocupados_ex"]/df_sexo["poblacion_total_ex"]

/tmp/ipython-input-2674306155.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_sexo= df_empleo.groupby("sexo")[["poblacion_total_ex","ocupados_ex"]].sum().reset_index()


In [490]:
# Celda de verificación (no modificar)
import numpy as np
dimensiones_esperadas = (2, 4)
assert df_sexo.shape == dimensiones_esperadas, \
    f"Las dimensiones son incorrectas. Se esperaba {dimensiones_esperadas}, pero se encontró {df_sexo.shape}. ¿Aplicaste .reset_index()?"
assert 'sexo' in df_sexo.columns, \
    "La columna 'sexo' no se encontró. ¿Olvidaste usar .reset_index() al final?"
assert not isinstance(df_sexo.index, pd.Index) or df_sexo.index.name != 'sexo', \
    "El índice del DataFrame sigue siendo 'sexo'. Asegúrate de haber aplicado .reset_index()."
print("¡Verificación completada con éxito! ✅")

¡Verificación completada con éxito! ✅


2. Crea un nuevo dataframe llamado "df_educacion" donde se agrupe por 'nivel_educativo' y se calcule la suma de las columnas 'poblacion_total_ex' y 'ocupados_ex'. Crea una nueva columna que se llame 'porcentaje_ocupados' donde dividas las columna 'ocupados_ex' sobre 'poblacion_total_ex'. Al final usa reset_index() para convertir las categorías de nivel educativo en una columna.

In [504]:
### TU CÓDIGO AQUÍ ###
df_educacion= df_empleo.groupby("nivel_educativo")[["poblacion_total_ex","ocupados_ex"]].sum().reset_index()
df_educacion["porcentaje_ocupados"]= df_educacion["ocupados_ex"]/df_educacion["poblacion_total_ex"]
orden_educativo = ['Ninguno', 'Educación media', 'Técnica profesional', 'Tecnológica', 'Universitaria', 'Especialización', 'Maestría', 'Doctorado','No sabe, no informa']
df_educacion["nivel_educativo"] = pd.Categorical(df_educacion["nivel_educativo"], categories=orden_educativo, ordered=True)
df_educacion = df_educacion.sort_values(by="nivel_educativo")

/tmp/ipython-input-2109698008.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_educacion= df_empleo.groupby("nivel_educativo")[["poblacion_total_ex","ocupados_ex"]].sum().reset_index()


In [492]:
# --- Celda de Verificación (No modificar) ---
import numpy as np
n_niveles_educativos = df_empleo['nivel_educativo'].nunique()
dimensiones_esperadas = (n_niveles_educativos, 4)
assert df_educacion.shape == dimensiones_esperadas, \
    f"Las dimensiones son incorrectas. Se esperaba {dimensiones_esperadas} (filas, columnas), pero se encontró {df_educacion.shape}."
columnas_esperadas = {'nivel_educativo', 'poblacion_total_ex', 'ocupados_ex', 'porcentaje_ocupados'}
assert set(df_educacion.columns) == columnas_esperadas, \
    f"Los nombres de las columnas no son correctos. Se esperaban {columnas_esperadas}."
assert 'nivel_educativo' in df_educacion.columns, \
    "La columna 'nivel_educativo' no se encontró. ¿Olvidaste usar .reset_index() al final?"
print("¡Verificación completada con éxito! ✅")

¡Verificación completada con éxito! ✅


In [505]:
display(df_sexo)
display(df_educacion)

sexo  poblacion_total_ex   ocupados_ex  porcentaje_ocupados
0  Hombre       25,255,111.00 13,797,363.08                 0.55
1   Mujer       26,682,244.83  9,707,731.65                 0.36

nivel_educativo  poblacion_total_ex  ocupados_ex  porcentaje_ocupados
4              Ninguno        1,710,908.13   348,733.89                 0.20
1      Educación media       14,327,525.06 8,605,757.92                 0.60
7  Técnica profesional        2,781,981.33 1,953,369.02                 0.70
6          Tecnológica        1,392,852.75 1,040,969.57                 0.75
8        Universitaria        3,731,910.34 2,759,724.79                 0.74
2      Especialización          965,236.30   777,759.47                 0.81
3             Maestría          448,666.05   386,964.59                 0.86
0            Doctorado           46,649.18    39,950.47                 0.86
5  No sabe, no informa               94.78        37.60                 0.40

3. En una celda de Markdown escribe qué conclusiones puedes sacar de la información obtenida en df_sexo y df_educacion.

Escribe tu análisis aquí: Por parte del dataframe de sexo se puede concluir que pese a que la población de mujeres sea mayor en comparacion a los hombres, estas tienen una menor ocupación y menor acceso al mercado laboral. Solo el 36% del total de mujeres tienen empleo mientras que por parte de los hombres el 55% del total de ellos cuentan con empleo. Con esto se puede evidenciar la brecha de genero existente en el mercado laboral que historicamente ha estado presente en Colombia.

Por parte dek dataframe de educacion, en términos generales, se puede concluir que a mayor nivel educativo, mayor es la probabilidad de estar empleado. El dataframe muestra un cambio significativo en el porcentaje de ocupación al pasar de no tener ningún tipo de estudio a alcanzar al menos la educación media. A partir de allí, se observa una tendencia creciente en el porcentaje de ocupación a medida que se avanza en los niveles educativos, alcanzando sus valores más altos en los niveles de especialización, maestría y doctorado.

Por ende, se concluye que el nivel educativo se consolida como un factor clave que incide de manera positiva y casi proporcional en la posibilidad de acceder a un empleo.

In [506]:
# Celda de verificación (no modificar)
print("\n¡Punto 9 completado con éxito! ✅")


¡Punto 9 completado con éxito! ✅


# Ejercicio 10: Exportar datos (5%)

Muchas veces requerimos exportar los resultados de nuestros análisis a Excel para compartir la información con otras personas.

Exporta los dataframes df_sexo y df_educacion a un Excel llamado 'Analisis_ocupados.xlsx'. Cada dataframe debe quedar en una hoja diferente. La hoja de 'df_sexo' se debe llamar 'sexo' y la hoja de 'df_educacion' se debe llamar 'educacion'.

In [507]:
### TU CÓDIGO AQUÍ ###
with pd.ExcelWriter('Analisis_ocupados.xlsx') as writer:
    df_sexo.to_excel(writer, sheet_name='sexo', index=False)
    df_educacion.to_excel(writer, sheet_name='educacion', index=False)

In [508]:
# Celda de verificación (no modificar)
import os
import pandas as pd
nombre_archivo = 'Analisis_ocupados.xlsx'
assert os.path.exists(nombre_archivo), \
    f"No se encontró el archivo '{nombre_archivo}'. ¿Lo exportaste con el nombre correcto?"
try:
    xls = pd.ExcelFile(nombre_archivo)
    nombres_hojas_encontrados = xls.sheet_names
    nombres_hojas_esperados = ['sexo', 'educacion']
    assert set(nombres_hojas_encontrados) == set(nombres_hojas_esperados), \
        f"Los nombres de las hojas no son correctos. Se esperaban {nombres_hojas_esperados} pero se encontraron {nombres_hojas_encontrados}."
    print("¡Verificación completada con éxito! ✅")
except Exception as e:
    print(f"Ocurrió un error al verificar el archivo: {e}")
print("\n¡Punto 10 completado con éxito! ✅")

¡Verificación completada con éxito! ✅

¡Punto 10 completado con éxito! ✅
